File: csv-tools/scripts-node/sort.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import { parseArgs } from "node:util";
import {
  DEFAULT_ENCODING,
  loadCsv,
  parseDelimiter,
  writeCsv,
  writeJson,
} from "./common.mjs";

function parseFields(raw) {
  return raw.split(",").map((f) => f.trim()).filter(Boolean);
}

function keyForRow(row, fields, numeric) {
  return fields.map((field) => {
    let v = (row[field] ?? "").trim();
    if (numeric && v) {
      const n = Number.parseFloat(v);
      if (!Number.isNaN(n)) return n;
    }
    return v || (numeric ? -Infinity : "");
  });
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      delimiter: { type: "string", default: "," },
      "no-header": { type: "boolean", default: false },
      "skip-lines": { type: "string" },
      "comment-char": { type: "string" },
      encoding: { type: "string", default: DEFAULT_ENCODING },
      by: { type: "string" },
      desc: { type: "boolean", default: false },
      numeric: { type: "boolean", default: false },
      format: { type: "string", default: "csv" },
      compact: { type: "boolean", default: false },
    },
    allowPositionals: true,
  });

  if (!values.by) {
    console.error("sort.mjs: --by is required");
    process.exit(1);
  }

  const input = positionals[0] ?? "-";
  const { columns, rows } = loadCsv(input, {
    delimiter: parseDelimiter(values.delimiter),
    hasHeader: !values["no-header"],
    commentChar: values["comment-char"] ?? null,
    encoding: values.encoding,
    skipLines: values["skip-lines"] != null ? Number.parseInt(values["skip-lines"], 10) : null,
  });

  let byFields = parseFields(values.by);
  byFields = byFields.filter((f) => columns.includes(f));
  if (!byFields.length) byFields = columns.slice(0, 1);

  const sorted = [...rows].sort((a, b) => {
    const ka = keyForRow(a, byFields, values.numeric);
    const kb = keyForRow(b, byFields, values.numeric);
    for (let i = 0; i < ka.length; i++) {
      const va = ka[i];
      const vb = kb[i];
      if (va < vb) return values.desc ? 1 : -1;
      if (va > vb) return values.desc ? -1 : 1;
    }
    return 0;
  });

  const delim = parseDelimiter(values.delimiter);
  if (values.format === "json") {
    writeJson(sorted, values.compact);
  } else {
    writeCsv(columns, sorted, delim, process.stdout);
  }
}

main();

File: csv-tools/scripts/sort.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Sort CSV rows by one or more columns."""

from __future__ import annotations

import argparse
import sys

from common import load_csv, parse_delimiter, write_csv, write_json


def add_dialect_args(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("input", nargs="?", default="-", help="Input CSV file or '-' for stdin.")
    parser.add_argument("--delimiter", default=",", help="Field delimiter (use '\\t' for tab).")
    parser.add_argument("--no-header", action="store_true", help="First row is data.")
    parser.add_argument("--skip-lines", type=int, default=None, help="Skip this many preamble lines; next line is header.")
    parser.add_argument("--comment-char", help="Skip lines starting with this character.")
    parser.add_argument("--encoding", default="utf-8", help="File encoding.")


def parse_fields(raw: str) -> list[str]:
    return [f.strip() for f in raw.split(",") if f.strip()]


def key_for_row(row: dict[str, str], fields: list[str], numeric: bool) -> tuple:
    def norm(field: str):
        v = (row.get(field, "") or "").strip()
        if numeric and v:
            try:
                return float(v)
            except ValueError:
                pass
        return v or (float("-inf") if numeric else "")

    return tuple(norm(f) for f in fields)


def main() -> None:
    parser = argparse.ArgumentParser(description="Sort CSV by columns.")
    add_dialect_args(parser)
    parser.add_argument("--by", required=True, help="Comma-separated columns to sort by.")
    parser.add_argument("--desc", action="store_true", help="Sort descending.")
    parser.add_argument("--numeric", action="store_true", help="Compare as numbers.")
    parser.add_argument("--format", choices=["csv", "json"], default="csv", help="Output format (default: csv).")
    parser.add_argument("--compact", action="store_true", help="Emit compact JSON when --format json.")
    args = parser.parse_args()

    columns, rows = load_csv(
        args.input,
        delimiter=parse_delimiter(args.delimiter),
        has_header=not args.no_header,
        comment_char=args.comment_char,
        encoding=args.encoding,
        skip_lines=args.skip_lines,
    )[:2]

    by_fields = parse_fields(args.by)
    by_fields = [f for f in by_fields if f in columns]
    if not by_fields:
        by_fields = columns[:1]

    sorted_rows = sorted(
        rows,
        key=lambda r: key_for_row(r, by_fields, args.numeric),
        reverse=args.desc,
    )

    if args.format == "csv":
        write_csv(columns, sorted_rows, parse_delimiter(args.delimiter), sys.stdout)
    else:
        write_json(sorted_rows, compact=args.compact)


if __name__ == "__main__":
    main()
